Possible fix for .NET 8 Blazor Web App breakage

WhitWaldo opened the pull request:
Adds .NET 8 as a valid target in addition to .NET 7 and the other versions.

Also moves state initialization out from `OnAfterRenderAsync` (which doesn't run on server anymore [per the docs](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/lifecycle?view=aspnetcore-8.0#after-component-render-onafterrenderasync)) to `OnInitializedAsync` which runs after existing logic in `OnInitialized` and after `OnParametersSet` (which in .NET 8 now runs before `OnInitialized{Async}`), so it's the last lifecycle method before `OnAfterRender`, assuming it were still being called.

I'm having problems getting Fluxor.Blazor.Web to build on my machine (likely missing one of your targets) so I haven't yet tested this, but as covered in the [related issue,](https://github.com/mrpmorris/Fluxor/issues/480), when I set a breakpoint in the ActionDispatched method, I can see that the actions are accumulating in the Store and they don't dequeue because the store never sets `HasActivatedStore` to true, which only happens via this call to `Store.InitializeAsync()` which only happens in `OnAfterRenderAsync` which isn't called on the server anymore, so it intuitively feels like this might fix the problem.

I'll follow up if I'm able to validate it locally furhter.

WhitWaldo commented:
There's unfortunately more that needs tackling here. I built a local version of Fluxor that only target .NET 8 to simplify this and it builds fine now, but while the Counter page loads without error when `@rendermode InteractiveServer` and I see the `DequeueActions` method trigger as expected, when I swap this out with `@rendermode InteractiveAuto`, now I'm getting a DI-related exception:

> Error: One or more errors occurred. (Cannot provide a value for property 'State' on type 'Blazor8Test.Client.Pages.Counter'. There is no registered service of type 'Fluxor.IState`1[Shared.State.CounterUseCase.CounterState]'.)

This is odd since I'm registering all dependencies on the primary project so it should flow through to the client (per the guidance [here](https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/dependency-injection?view=aspnetcore-8.0#client-side-services-that-fail-during-prerendering)) but this remains a work in progress.

mrpmorris commented:
Fancy pairing on this one if the days?

I'm available from 5PM week days (UK time)

WhitWaldo commented:
Sure - we can sync on that next week. I'm in the US, but if I did the math right, that puts me at 11 AM or later, so that's more than fine.

After further experimenting, I found that the DI problem is resolved by disabling prerendering (putting `@rendermode @(new InteractiveAutoRenderMode(prerender:false))` up at the top of the Counter.razor in the wasm project). Now we're back at the original problem though in that I click the button and the state isn't initializing (the `QueuedActions` fills with button clicks, but nothing otherwise happens).

WhitWaldo commented:
Ok, I'm thoroughly unaware as to why, but on a whim I changed the default option to use a singleton for the store lifetime (since that's what it's using on the client anyway as scoped isn't supported) and the application worked end to end.. until I left the page and came back to it. At that point, DI failed and indicated it couldn't resolve IState<CounterState> despite it having just worked before.

I'm going to have to figure out how one knows when it's transitioned from server to client as I suspect that the first time I loaded it, it was running on the server and working as expected. I gather that the transition to the wasm rendering mode happens at a page change, so when I went to another page and came back, presumably I was now running on the wasm project which led to this not working.

Of course that doesn't make a ton of sense as anything that's registered on the server _should_ be available in the client project as well, just not necessarily vice-versa. As I'm fully registering Fluxor on the server, everything should be available for wasm, so it's unclear why that isn't happening.

WhitWaldo commented:
I've added a single persistence layer to Fluxor that appears to resolve the cross-rendermode issue. As observed by others on another thread about this, it's one thing to get Fluxor up and running again, but a whole other thing to completely lose your state as soon as you jump render modes (e.g. server -> web assembly). 

While I've taken a stab at implementing this in a lightweight manner, it wasn't until I was integrating the changes from my test project to this branch that I realized my solution is only compatible with .NET 6 and higher because of my use of `JsonNode` in `System.Text.Json`, so I surrounded the relevant registration and implementation bits with conditional compiler flags in the shorter term. Also, I cannot get Fluxor to build on my machine as it's targeting frameworks I simply don't have installed. If you've got .NET 8 installed, clone my repo below for a .NET 8-only version of this.

A .NET 8 version of this can be seen on the [latest commit here](https://github.com/WhitWaldo/Blazor8FluxorTest/tree/persistence), but I'll take a moment to walk through the idea.

I built off the idea of using an action to identify when the state has initialized and an effect that triggers from that. I've added some extensions to the registration method that allows the developer to opt into the persistence feature and provide an implementation of `IPersistenceManager` (I wrote such an implementation in my other project using session storage [here](https://github.com/WhitWaldo/Blazor8FluxorTest/blob/persistence/Shared/Persistence/SessionStorePersistenceManager.cs)) and to allow them to register any additional dependencies here that it might require. If `IPersistenceManager` is never registered, Fluxor runs as it did before this implementation and that's it (with a few minor tweaks).

If it _is_ registered, when Fluxor is activating the store instance, it first marks the store as activated, dequeues any actions and then dispatches a new `StoreRehydratingAction` that kicks off a registered effect that reads any persisted state from the `IPersistenceManager` and uses a new method on the store to rehyrdrate the state on a per-feature basis (using that `RestoreState` method that explicitly says it shouldn't be used for this - whoops). This concludes with a `StoreRehydratedAction` in case anyone wants to subscribe to it for some reason.

On a location change event in the Fluxor component, it dispatches a `PersistingStoreAction` and this does the opposite. It builds a JSON object (each feature to a separate named key) and persists this via the `IPersistenceManager` and fires off a `PersistedStoreAction`.

This implementation avoids having the project take on any other package dependencies and leaves it to the developer to decide where and how they want to persist any data, whether that be in session state as my demo does or bind it to a session ID and save it somewhere on the server - their choice. Just implement the `IPersistenceManager` and register with the provided method on `FluxorOptions`.

Remaining work:
- Again, this is only compatible with .NET 6, .NET 7 and .NET 8 because of the use of System.Text.Json for building the JsonNode during the serialization and deserialization steps. I imagine the same can be done with Newtonsoft.Json, if you want to take on another dependency.
- I see some flicker in WebAssembly where it's initializing the store, then reading the state and updating it, so this could be improved on.
- I thought persisting it on location change is often enough (wouldn't want to do it every single action) as I don't know that Blazor jumps render mode until you actually change the page. In all my testing, it always required a page change to swap modes. If that changes in a future release, perhaps they'll give us an event to help out. But if there's some notification of this or users want to save more frequently, this approach will need to be tweaked.
- I save to the browser's session store in my demo so that when I relaunch the app, it doesn't just load the state from the previous session. I wasn't able to find any sort of common value (for non-authenticated users) that give me some sort of identifier indicating that WASM-session X matched Server-session X (so that if X didn't match, it would trigger a clear of the store). I added a method to `IPersistenceManager` so this is covered in a server-side store scenario, so perhaps this can be improved on.

GordonBeeming commented:
@WhitWaldo thanks for the effort in getting this working  ❤️ 

@mrpmorris 
1. Would this fix make it into 5.9.x or only 6.0?
2. If not, do you know roughly when you are wanting 6.0 to reach stable?

jafin commented on Source/Lib/Fluxor/Fluxor.csproj:
```diff
@@ -15,9 +15,11 @@
 		<PackageReference Condition="'$(TargetFramework)' == 'net7.0'" Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="7.0.0" />
 		<PackageReference Condition="'$(TargetFramework)' == 'net6.0'" Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="6.0.0" />
 		<PackageReference Condition="'$(TargetFramework)' == 'net5.0'" Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="5.0.0" />
 		<PackageReference Condition="'$(TargetFramework)' != 'net7.0' AND '$(TargetFramework)' != 'net6.0' AND '$(TargetFramework)' != 'net5.0'" Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="3.1.0" />
```
Does this need to be removed or a condition added for `!= 'net8.0` ?

WhitWaldo replied:
I rather assumed that whatever was there was compatible up to the current supported build prior to my adding .NET 8 to the list - presumably that includes .NET Core 3.1 as-is if that's still a supported target, so I wouldn't think anything else is necessary.

jafin commented on Source/Lib/Fluxor/Store.cs:
```diff
@@ -14,6 +21,10 @@ public class Store : IStore, IActionSubscriber, IDisposable
 		public Task Initialized => InitializedCompletionSource.Task;
 
+#if NET6_0_OR_GREATER
+		private readonly IPersistenceManager? _persistenceManager;
```
NIT: Consider following naming convention of project. Uppercase vs leading underscore for privates.

WhitWaldo replied:
I defer to the maintainer's preference here.

mikeppcom commented:
Will this PR be in beta#3/V6? We make heavy use of Fluxor and are moving to a hybrid Blazor server/client model and are facing the issue that actions are no longer being dispatched.

wave9d commented:
> Will this PR be in beta#3/V6? We make heavy use of Fluxor and are moving to a .NET8 hybrid Blazor server/client render model and are facing the issue that actions are no longer being dispatched.

Any news on this PR being available for a beta release ? We're in the same situation as @mikeppcom

mrpmorris commented:
try specifying the @rendermode in `Routes.razor` or on the `<StoreInitializer` compoonent

orosbogdan commented:
> try specifying the @rendermode in `Routes.razor` or on the `<StoreInitializer` compoonent

The problem would be if you wanted to have both server interactive and wasm interactive pages in your app right? 
Same for auto render mode I think.

WhitWaldo commented:
> > try specifying the @rendermode in `Routes.razor` or on the `<StoreInitializer` compoonent
> 
> The problem would be if you wanted to have both server interactive and wasm interactive pages in your app right? Same for auto render mode I think.

If you don't specify an interactive mode (e.g. InteractiveServer, InteractiveWebAssembly or InteractiveAuto), you simply won't see any changes made to the page/component. If the changes are in response to interaction with an HTML element, the method isn't called and thus the value isn't updated and propagated and no changes are made to the page. For any interactivity, you have to specify one of the above interactive modes.

If your whole app already uses strictly interactive WASM or interactive server, I don't know that you'd have a problem with Fluxor's latest public build. This PR was made specifically to support InteractiveAuto as it's the transition from Server to WebAssembly (and back) that was causing issues as the state wouldn't be available across contexts.

orosbogdan commented:
> > > try specifying the @rendermode in `Routes.razor` or on the `<StoreInitializer` compoonent
> > 
> > 
> > The problem would be if you wanted to have both server interactive and wasm interactive pages in your app right? Same for auto render mode I think.
> 
> If you don't specify an interactive mode (e.g. InteractiveServer, InteractiveWebAssembly or InteractiveAuto), you simply won't see any changes made to the page/component. If the changes are in response to interaction with an HTML element, the method isn't called and thus the value isn't updated and propagated and no changes are made to the page. For any interactivity, you have to specify one of the above interactive modes.
> 
> If your whole app already uses strictly interactive WASM or interactive server, I don't know that you'd have a problem with Fluxor's latest public build. This PR was made specifically to support InteractiveAuto as it's the transition from Server to WebAssembly (and back) that was causing issues as the state wouldn't be available across contexts.

Would an blazor project having 2 separate pages, 1 server interactive and 1 wasm interactive work with latest fluxor build ?

WhitWaldo commented:
> > > > try specifying the @rendermode in `Routes.razor` or on the `<StoreInitializer` compoonent
> > > 
> > > 
> > > The problem would be if you wanted to have both server interactive and wasm interactive pages in your app right? Same for auto render mode I think.
> > 
> > 
> > If you don't specify an interactive mode (e.g. InteractiveServer, InteractiveWebAssembly or InteractiveAuto), you simply won't see any changes made to the page/component. If the changes are in response to interaction with an HTML element, the method isn't called and thus the value isn't updated and propagated and no changes are made to the page. For any interactivity, you have to specify one of the above interactive modes.
> > If your whole app already uses strictly interactive WASM or interactive server, I don't know that you'd have a problem with Fluxor's latest public build. This PR was made specifically to support InteractiveAuto as it's the transition from Server to WebAssembly (and back) that was causing issues as the state wouldn't be available across contexts.
> 
> Would an blazor project having 2 separate pages, 1 server interactive and 1 wasm interactive work with latest fluxor build ?

If they use separate state from each other, the latest public build should work fine. Your issue is going to if you attempt to mix state between the two because they have no mechanism to communicate it back and forth. 

That's the point of this PR, is to add a mechanism to persist the state regardless of where it's being updated so that either environment can access the latest version of that state, even if any given component is shifting between the two (e.g. per InteractiveAuto it'll start running on the server and eventually run from WASM). 

The idea is that the data is persisted on the server via an `IPersistenceManager` ([my demo](https://github.com/WhitWaldo/Blazor8FluxorTest/tree/persistence) simulates a Redis instance by using an in-memory key/value store as a singleton service) and then implement an `IStateService` on both the server and wasm clients to persist and retrieve the state. On the server, this pulls from the `IPersistenceManager`, on the client, this implements an `HttpClient` to get the same from the server.

The idea is that you add an `IStateService` and a slightly different implementation for it whether it's running on the server or web assembly. As the state is updated through Fluxor, it'll be persisted and retrieved through this `IStateService` ensuring that it's all in sync regardless of whether the next component accessing it is doing so from the client or server.

jafin commented:
This didn't make it into v6?

mikeppcom commented:
Jason,

There isn’t an issue with Fluxor as such. If you’re using it with a web app you need to initialize it in the routes component with a render mode.

<Fluxor.Blazor.Web.StoreInitializer @rendermode="InteractiveServer" />

Thanks to Pete for pointing this out.

By the way not sure that webapps are really production ready as the only way I’ve been able to get them to work with layouts is to set the render mode at a global level and switch between server and client using a hard navigation.

Michael


From: Jason Finch ***@***.***>
Sent: Monday, June 10, 2024 6:01 AM
To: mrpmorris/Fluxor ***@***.***>
Cc: Michael Aston ***@***.***>; Mention ***@***.***>
Subject: Re: [mrpmorris/Fluxor] Possible fix for .NET 8 Blazor Web App breakage (PR #481)

EXTERNAL E-MAIL – This email has come from an external source. Verify the sender before responding, clicking on any links or opening attachments. Report suspicious emails.
________________________________
This didn't make it into v6?

—
Reply to this email directly, view it on GitHub<https://github.com/mrpmorris/Fluxor/pull/481#issuecomment-2157272642>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/A6CYAQJDGHV4ODVTL6RWPG3ZGUXKJAVCNFSM6AAAAABEMHGLVSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDCNJXGI3TENRUGI>.
You are receiving this because you were mentioned.Message ID: ***@***.******@***.***>>
______________________________________________________________________________________________

This e-mail message (including any attachments) from PayPoint plc or any member of the PayPoint 
group of trading companies (collectively referred to as 'PayPoint') is confidential and for the 
personal use of the recipient(s) identified above. The message may also be legally privileged. 
If it is not intended for you, do not disclose, copy or distribute the message but delete it 
immediately and notify the sender.

Any views or opinions expressed in this message are those of the author and not necessarily 
those of PayPoint. The message shall not form part of any legally binding contract or 
obligation.

PayPoint electronic communication systems are monitored without user consent to: 
investigate or to detect unauthorised use; prevent or detect crime; establish the existence 
of facts relevant to PayPoint; or ascertain compliance with regulatory practices relevant 
to the business of PayPoint.

This e-mail has been scanned for viruses by a third party e-mail management provider. PayPoint
cannot accept liability for any damage which you may suffer as a result of virus infection. We
recommend that you carry out your own virus checks before opening any attachment.
 
PayPoint plc – Registered in England number 3581541
Registered office: 1 The Boulevard, Shire Park, Welwyn Garden City, Hertfordshire, AL7 1EL
______________________________________________________________________________________________

WhitWaldo commented:
> Jason, There isn’t an issue with Fluxor as such. If you’re using it with a web app you need to initialize it in the routes component with a render mode. <Fluxor.Blazor.Web.StoreInitializer @rendermode="InteractiveServer" /> Thanks to Pete for pointing this out. By the way not sure that webapps are really production ready as the only way I’ve been able to get them to work with layouts is to set the render mode at a global level and switch between server and client using a hard navigation. 

You're right in that Fluxor will work fine so long as your application and components run entirely in either a server (personally tested) or web assembly (untested) context. Sure, if you have components that run on WebAssembly but don't inject Fluxor state, they'll work fine too. But merely running initializing the store on the server isn't going to suddenly open up some backchannel for your WebAssembly components that take a Fluxor state dependency to communicate and keep in sync with it.

Rather, this PR adds a persistence layer using the same approach provided in the sparse guidance for using the new Web Apps and leaves it as an exercise for the developer to decide just how the state itself is persisted (on the server) and how the WebAssembly components will communicate with the server to read and write its own updates. When the component is initialized then, it queries this central store (which is accessible now regardless of where the component is running) and when it applies the batch of changes, it persists that result back (so other components running in either context can access the same).
